### PR TITLE
Gallery Template

### PR DESCRIPTION
**Pull request by noslouch**
It's another big one! Will try to rebase some of these commits onto wagtail to try and scope it a bit.

First thing to note is that this is diff against the article template PR #76 because articles and galleries are pretty tightly coordinated.

### notes

- [the `GALLERY_PATH` const](https://github.com/nypublicradio/gothamist-web-client/pull/77/commits/158d2f92f8cf9fa34ffa04c3fe64cbbdeaa8b144) [is used to construct a gallery's url](https://github.com/nypublicradio/gothamist-web-client/pull/77/commits/2783eb09f3982ed8e487c32d980d6c3cc4b593e7). It's expected that all galleries follow the convention `<section slug>/galleries/<gallery slug>`, but when we release, that's going to be `<section slug>/photos/<gallery slug>`, and I only want to update that value in one place
- there wasn't an acceptance test for navigating into and out of galleries previously, but I've added one here
- I was able to move a lot of look up logic into the application adapter and [serializer](https://github.com/nypublicradio/gothamist-web-client/pull/77/commits/8a01463f3305a25cbb6b177d7bd3972a9b572908), which keeps the model-specific classes free from boilerplate
- some mirage values are only meant for the server, so [I added a `cleanMirageAttrs` function to enforce a blacklist](https://github.com/nypublicradio/gothamist-web-client/pull/77/commits/6133aed8776752f9af39a6e6815c10b485527dab). I don't love the api for this (it'd be nice if you could set a blacklist options on the mirage serializer itself), but it works for now
- wagtail serializes gallery slides into a deeply nested object, so [I'm flattening it out a bit int the gallery serializer on the client side](https://github.com/nypublicradio/gothamist-web-client/pull/77/commits/7373011a66ad7b930061e3a6480d0b8983faa4db)

### custom slide dimensions

I needed a way to specify slide sizes for different contexts: the gallery lead component on article templates (preview image and thumbnail) and the gallery slides themselves. The logic for each specific instance was relatively straight forward: map over the slides and produce a hash of sizes using the `wagtail-image-url` helper, but it occurred to me that rewriting the logic every time is going to be a hassle. 

[I added a `makeSizes` method to the gallery model](https://github.com/nypublicradio/gothamist-web-client/pull/77/commits/2927d4a6a6096116ffd92a0836b05427c76a0886) ([doc string here](https://github.com/nypublicradio/gothamist-web-client/pull/77/commits/3986ba28fdeaec1c9f6e6cabe81c1d90eadc68c7))that takes a hash of arbitrary size names and size parameters and returns an array of objects that can be passed into the various gallery contexts. The upside is the logic for this is encapsulated in one place, but the downside is the model now has a dependency on a third-party library, `ember-wagtail-images` (second-party, really, b/c it's our own lib). I'm open to suggestions on this one.

### relationships
I opted to make galleries and articles related to each other, but it might wind up being more trouble than it's worth, at least keeping them async. The upside is we can leverage the framework to pull in the data for us, the downside is fastboot doesn't seem to be totally aware of async relationships inside model hooks, because I kept seeing warnings locally.

- some nice ways to configure the relationships between articles and galleries
  - https://github.com/nypublicradio/gothamist-web-client/pull/77/commits/99ee6a9ab670c46a3a2ce7251addf78fe2b46e16
  - https://github.com/nypublicradio/gothamist-web-client/pull/77/commits/a29caea782dce1b32479a2ca6424494c9d27cd44
- work arounds for fastboot async leaks
  - https://github.com/nypublicradio/gothamist-web-client/pull/77/commits/d06f5edeced897f111b018b963ef5c6a51b4053d
  - https://github.com/nypublicradio/gothamist-web-client/pull/77/commits/f52b5add874d8adcce8d6b01ba006c895be1802c
  - https://github.com/nypublicradio/gothamist-web-client/pull/77/commits/e554d25833f3e34e191a4a7ab54b820531bff98e

[ticket](https://jira.wnyc.org/browse/CMS-547)

depends on nypublicradio/nypr-design-system#30

### CODE REVIEW

File: app/models/gallery.js [hunk 1]
```diff
@@ -1,6 +1,108 @@
...
+    @return {Array[Object]}
+  */
+  makeSizes(ops = {}, others = {}) {
+    return this.slides.map(slide => {
+      let sizes = Object.keys(ops)
+        .reduce((o, key) => ({
+          ...o,
+          [key]: wagtailImageUrl([slide.image, ...deriveWagtailArgs(ops[key])]),
+        }), {})
+      // sizes = Object.fromEntries(sizes);
```
Comment (mike-hearn) [inline app/models/gallery.js]: is this an errant comment or an explanation for what these lines do?

File: app/models/gallery.js [hunk 2]
```diff
@@ -1,6 +1,108 @@
...
+const CONTAIN = 'contain';
+const FITS = [FILL, COVER, CONTAIN];
+
+export default Page.extend({
+  // slides are normalized from the server to look like so:
+  // {
+  //   title String,
+  //   image Object{id Number, caption String},
+  // }
+  slides: DS.attr({deafultValue: () => []}),
```
Comment (walsh9) [inline app/models/gallery.js]: default is spelled wrong here

File: app/models/page.js [hunk 3]
```diff
@@ -19,10 +20,12 @@ export default DS.Model.extend({
...
 
-  // computed
+  // computeds
+  slug: reads('meta.slug'),
+
   path: computed('meta.html_url', function() {
-    if (typeof this.meta.html_url === 'string') {
-      return this.meta.html_url.replace(/https?:\/\/[^/]+\//, '');
+    if (this.meta && typeof this.meta.html_url === 'string') {
+      return this.meta.html_url.replace(/https?:\/\/[^/]+\//, '').replace(/\/$/, '');
```
Comment (mike-hearn) [inline app/models/page.js]: if we re-use this double regex pattern in any places might be worth putting into a separate function (not necessary if it's just a one-off, but wanted to flag it)

File: app/templates/gallery.hbs [hunk 4]
```diff
@@ -1,12 +1,14 @@
 {{set-body-class "gallery"}}
 <DoTargeting key='Template' value='Article Gallery' />
 <NyprOGalleryOverlay
-  @title={{model.title}}
-  @parentRoute={{array 'article' model.path}}
-  @slides={{model.gallery.slides}}
+  @title={{parentArticle.title}}
+  @parentRoute={{array 'article' parentArticle.section.slug parentArticle.path}}
+  @slides={{this.slides}}
```
Comment (mike-hearn) [inline app/templates/gallery.hbs]: just curious, any reason we use `this` with `this.slides` but not with `parentArticle`?

File: package.json [hunk 5]
```diff
@@ -85,7 +85,7 @@
     "memory-scroll": "^0.9.1",
     "nypr-ads": "^1.1.1",
     "nypr-auth": "^0.2.4",
-    "nypr-design-system": "^1.1.3",
+    "nypr-design-system": "nypublicradio/nypr-design-system#brian/cms-547",
```
Comment (mike-hearn) [inline package.json]: don't forget to update this before merge

File: app/models/gallery.js (hunk 1 above)
Comment (walsh9) [reply app/models/gallery.js]: I ran into a fastboot problem with node support for `Object.fromEntries` when trying to use it for some serialization.

But if we do want to use `fromEntries`  theres some very small polyfill code we could use, or even just add as a `utils` file we could import from: https://github.com/feross/fromentries/blob/master/index.js

File: app/models/gallery.js [hunk 6]
```diff
@@ -1,6 +1,108 @@
...
+
+  /**
+    Generate a map of arbitrary keys to image urls specified by parameters.
+
+    Values of the `ops` param will be interpreted in the following way:
+    - Number - resize width according to given value
+    - Array[size Number, dimension String('height'|'width')] - resize to given `size` along given `dimension` string
+    - Array[size Number, fit String('fill'|'cover'|'contain')] - crop to a square according to `size`, and resize according to `fit`
+    - Array[width Number, height Number, fit String('fill'|'cover'|'contain')] - crop to give `width` and `height` according to given `fit`
+
```
Comment (walsh9) [inline app/models/gallery.js]: It feels complicated that there are 4 different ways (or more?*) to define sizes with `makeSizes`. One is even a special shorthand for creating squares.

*(for example not sure how `thumb: [150, 150]` fits into the options defined in the doc string here. And it's not using the square shorthand here, is it doing something different or not?)

Anyway, I'm not sure the reduction in verboseness when calling the function is worth the increased complexity. What if all three arguments were required all the time? Or if kept optional, what about using named options in an object? Not sure...

File: app/models/gallery.js [hunk 7]
```diff
@@ -1,6 +1,108 @@
...
+          [key]: wagtailImageUrl([slide.image, ...deriveWagtailArgs(ops[key])]),
+        }), {})
+      // sizes = Object.fromEntries(sizes);
+      return {
+        title: slide.title,
+        caption: slide.image.caption,
+        ...sizes,
+        ...others,
+      };
+    })
```
Comment (walsh9) [inline app/models/gallery.js]: I'm also not sure makeSizes needs to be responsible for adding title and caption and optional arbitrary parameters.  What if it just returned sizes for one slide and was called like this? (Based off the gallery controller)


```js
slides: computed('model.gallery', function() {
  // make images for each breakpoint
  return this.model.gallery.slides.map((slide) => {
    title: slide.title,
    caption: slide.image.caption,
    ...this.model.gallery.makeSizes(slide, {
      thumb: [150, 150, 'fit'],
      srcS: [420, null, 'width'],
      srcM: [800, null, 'width'],
      srcL: [1200, null, 'width'],
    })
    width: 1200, //this is `...others`
  })
});
```

Or if we want to leave certain parameters optional, what about using named parameters?

```js
  thumb: {width: 150, width: 150},
  srcS: {width: 420},
  srcM: {width: 800},
  srcL: {width: 1200},
```

Or do we need makeSizes at all?
```js
return this.model.gallery.slides.map((slide) => {
  title: slide.title,
  caption: slide.image.caption,
  width: 1200,
  thumb: wagtailImageUrl([slide.image, 150, 150]),
  srcS: wagtailImageUrl([slide.image, 420]),
  srcM: wagtailImageUrl([slide.image, 800]),
  srcL: wagtailImageUrl([slide.image, 1200]),
});
```

Just throwing some ideas out here.

File: app/models/gallery.js (hunk 1 above)
Comment (noslouch) [reply app/models/gallery.js]: yeah I encountered that node issue, but I'm going to rethink `makeSizes` based on @walsh9's feedback

File: app/templates/gallery.hbs (hunk 4 above)
Comment (noslouch) [reply app/templates/gallery.hbs]: no reason. I believe `this` should only be used in components tho. will update.

File: app/models/gallery.js (hunk 7 above)
Comment (noslouch) [reply app/models/gallery.js]: yeah, I think I was a little overzealous with slide generation. I removed `makeSizes` entirely. Thanks for the note!

File: app/models/gallery.js (hunk 6 above)
Comment (noslouch) [reply app/models/gallery.js]: agree!

**Comment by noslouch** [comment]
the base branch was merged into the integration branch so I'm going to close this in favor of a new PR based on `wagtail`

**Pull request closed without merging**